### PR TITLE
CCEventDispatcher crash fix

### PR DESCRIPTION
**Pull request by BodbDearg**
Fix possible crashes which could be caused by the EventDispatcher having listeners associated with nodes are destroyed. Catch the case where a node registers a listener while we are dispatching events and gets destroyed while that event is still being dispatched. Check the list of nodes to be added into the event dispatcher fully when we are cleaning listeners for a target node.

This issue was found using the extra debug verification in this pull request:
https://github.com/cocos2d/cocos2d-x/pull/6011

### CODE REVIEW

**Comment by BodbDearg** [comment]
Just added a test which can be used to reproduce the crashing issue and verify the fix.

**Comment by dumganhar** [comment]
Thank you. I'm checking this issue. :)

File: cocos/2d/CCEventDispatcher.cpp [hunk 1]
```diff
@@ -340,6 +340,26 @@ void EventDispatcher::removeEventListenersForTarget(Node* target, bool recursive
...
+    // This is to catch the scenario where the node gets destroyed before it's listener
+    // is added into the event dispatcher fully. This could happen if a node registers a listener
+    // and gets destroyed while we are dispatching an event (touch etc.)
+    for (auto iter = _toAddedListeners.begin(); iter != _toAddedListeners.end(); )
+    {
+        EventListener * listener = *iter;
+            
+        if (listener->getSceneGraphPriority() == target)
+        {
+            listener->release();
```
Comment (dumganhar) [inline cocos/2d/CCEventDispatcher.cpp:354]: It's better to invoke

``` c++
listener->setRegistered(false);
```

before

``` c++
listener->release();
```

Comment (dumganhar) [inline cocos/2d/CCEventDispatcher.cpp:354]: Since the listener may be reused after it was removed.

**Comment by dumganhar** [comment]
Merged at https://github.com/cocos2d/cocos2d-x/pull/6162 . Thanks.:)

**Comment by ricardoquesada** [comment]
A few days ago I decided to try cpp-test on Win32 and it crashes all the time on the `EventDispatcher`

When I saw this patch, I thought it was refering to the win32 bug but now... `cpp-test` is still crashing with `cpp-test`.

any idea? this bug is a show-stopper for v3.0.

How to reproduce it:
- run cpp-test on windows
- compile it using Visual Studio 2013 Update 2

cpp-empty-test works ok. But cpp-test crashes when I launch the app.

**Comment by dumganhar** [comment]
@ricardoquesada , thanks for your feedback. I will check it ASAP. :)

File: cocos/2d/CCEventDispatcher.cpp (hunk 1 above)
Comment (BodbDearg) [inline cocos/2d/CCEventDispatcher.cpp:354]: Great point, thanks for catching this!

**Comment by BodbDearg** [comment]
That's very concerning, thanks for letting us know! I think I'll check this out too.

**Comment by dumganhar** [comment]
@ricardoquesada  @DarraghCoy 
I fixed it at https://github.com/cocos2d/cocos2d-x/pull/6169

**WHY DOES THE ISSUE ONLY EXIST IN cpp-tests on win32:**

cpp-tests registers `Mouse listener`, and if we launch application with mouse moving frequently, 
the mouse listener will receive mouse event before the first scene is ran.
So it will crash at `sortSceneGraphPriorityListeners` since `getRunningScene` will return nullptr.

Why don't other desktop platforms (like Mac or Linux) have this issue?

I think it's because different implementation in glfw.

**Comment by ricardoquesada** [comment]
@dumganhar 
good finding! thanks.

**Comment by BodbDearg** [comment]
Awesome, thanks for fixing up that issue!

**Comment by dumganhar** [comment]
My pleasure. :)

**Pull request merged**